### PR TITLE
Enable recording hook calls

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Enable recording hook calls. In this PR I add a SQL table to store each call to a hook. These logs can be monitored inside the VueJS APP in BO.
| Type?         | new feature
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/ps_qualityassurance/issues/6
| How to test?  | 

It looks like this:
Capture d’écran 2020-12-03 à 11 28 55
Capture d’écran 2020-12-03 à 11 29 02

So basically
- every time you use ps_qualityassurance connected to one hook
- this call is recorded in a logs table
- logs table can be viewed in the Module BO page

So you can see what hooks have been called, with date, parameters, output :)

### CODE REVIEW

File: ps_qualityassurance.php [hunk 1]
```diff
@@ -68,7 +76,20 @@ public function install()
...
+            '`request_identifier` varchar(255) NOT NULL,' .
+            '`hook_name` varchar(255) NOT NULL,' .
+            '`hook_parameters` text NOT NULL,' .
+            '`output` text NOT NULL,' .
+            '`called_at` datetime NOT NULL,' .
+            '`error` TINYINT(1) NOT NULL,' .
+            'PRIMARY KEY (`id`)' .
+            ') ENGINE=' . _MYSQL_ENGINE_ . ' DEFAULT CHARSET=UTF8;';
+
+        return Db::getInstance()->execute($query1)
```
Comment (Progi1984) [inline ps_qualityassurance.php:90]:
```suggestion
        return Db::getInstance()->execute($query)
```

File: ps_qualityassurance.php [hunk 2]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
     {
         $hookName = preg_replace('~^hook~', '', $methodName);
 
+        $arguments[4] = ['4' => ['5' => []]];
```
Comment (PierreRambaud) [inline ps_qualityassurance.php:105]: I don't understand this line :thinking:

File: ps_qualityassurance.php [hunk 3]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
+            $this->recordHookCall($hookName, $params, $output, true);
+
+            echo $output;
+        }
+    }
+
+    public function getContent()
+    {
+        $moduleAdminLink = Context::getContext()->link->getAdminLink('AdminQualityAssurance', true);
+        Tools::redirectAdmin($moduleAdminLink);
```
Comment (Progi1984) [inline ps_qualityassurance.php:131]:
```suggestion
        Tools::redirectAdmin(Context::getContext()->link->getAdminLink('AdminQualityAssurance', true););
```

File: ps_qualityassurance.php [hunk 4]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
+    }
+
+    /**
+     * Look whether there is a payload registered for this hook
+     *
+     * @param string $hookName
+     * @return array|null
+     */
+    protected function getRegisteredHookPayload($hookName)
```
Comment (Progi1984) [inline ps_qualityassurance.php:140]:
```suggestion
    protected function getRegisteredHookPayload($hookName): ?array
```

File: ps_qualityassurance.php [hunk 5]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
+
+    /**
+     * @param string $hookName
+     * @param array $arguments
+     * @param string|null $output
+     * @param bool|null $error
+     *
+     * @return bool
+     */
+    protected function recordHookCall(string $hookName, array $arguments, ?string $output = null, ?bool $error = false)
```
Comment (Progi1984) [inline ps_qualityassurance.php:161]:
```suggestion
    protected function recordHookCall(string $hookName, array $arguments, ?string $output = null, ?bool $error = false): bool
```

Comment (PierreRambaud) [inline ps_qualityassurance.php:161]: Nope, it breaks the compatibility with PS 1.7.1. Keep the 5.6 compatiblity

File: ps_qualityassurance.php [hunk 6]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
+
+        return Db::getInstance()->execute($sql);
+    }
+
+    /**
+     * @param array $arguments
+     *
+     * @return string
+     */
+    protected function makeArgumentsReadable(array $arguments)
```
Comment (Progi1984) [inline ps_qualityassurance.php:187]:
```suggestion
    protected function makeArgumentsReadable(array $arguments): string
```

File: ps_qualityassurance.php [hunk 7]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
         }
+
+        return json_encode($result);
     }
 
-    public function getContent()
+    /**
+     * @return string
+     */
+    protected function getRequestIdentifier()
```
Comment (Progi1984) [inline ps_qualityassurance.php:206]:
```suggestion
    protected function getRequestIdentifier(): string
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: ps_qualityassurance.php (hunk 2 above)
Comment (matks) [reply ps_qualityassurance.php:105]: Whoops!

Debug code written at 10p 😅  indeed should not be there

File: ps_qualityassurance.php [hunk 8]
```diff
@@ -81,26 +102,113 @@ public function __call($methodName, array $arguments)
...
     {
-        $moduleAdminLink = Context::getContext()->link->getAdminLink('AdminQualityAssurance', true);
-        Tools::redirectAdmin($moduleAdminLink);
+        if ($this->requestIdentifier === null) {
+            $this->requestIdentifier = random_bytes(20);
```
Comment (PierreRambaud) [inline ps_qualityassurance.php:209]: I don't see why you use a requestIdentifier when you have an id in the table :thinking:

Comment (matks) [reply ps_qualityassurance.php:209]: I want to group together hook calls performed in the same request->response process.
So if a HTTP request hits  `/homepage` endpoint and it triggers 4 hook calls I want to be able to connect these 4 hook calls together. This helps the QA analyst to see "all hook calls triggered by my latest interaction with PrestaShop".

File: ps_qualityassurance.php (hunk 5 above)
Comment (matks) [reply ps_qualityassurance.php:161]: Yes. Initially I wanted php5.6 compatibility but I added this typing mechanically 😅  just habits.

Comment (matks) [reply ps_qualityassurance.php:161]: As mentioned by @PierreRambaud let's try to keep php5.6 compatibility here

File: ps_qualityassurance.php (hunk 6 above)
Comment (matks) [reply ps_qualityassurance.php:187]: As mentioned by @PierreRambaud let's try to keep php5.6 compatibility here

File: ps_qualityassurance.php (hunk 4 above)
Comment (matks) [reply ps_qualityassurance.php:140]: As mentioned by @PierreRambaud let's try to keep php5.6 compatibility here

File: ps_qualityassurance.php (hunk 3 above)
Comment (matks) [reply ps_qualityassurance.php:131]: Out of the scope of this PR I did not modify this line (git thinks it was moved but it was not)

By the way 😄  I really dislike calling 4 functions on the same line, I find it very unreadable and it does not help when using break points.
You'll see this debate is very open with people on each side https://stackoverflow.com/questions/18250389/nested-function-calls-whats-the-best-practice

File: ps_qualityassurance.php (hunk 8 above)
Comment (PierreRambaud) [reply ps_qualityassurance.php:209]: Oh, I understand now. But prever uniqid() instead of random_bytes :)

Comment (matks) [reply ps_qualityassurance.php]: Fixed in https://github.com/PrestaShop/ps_qualityassurance/pull/8/commits/da03ba8a722b9e6ec2da896b72f65667ddc7c0a4

File: ps_qualityassurance.php (hunk 5 above)
Comment (matks) [reply ps_qualityassurance.php]: Fixed in https://github.com/PrestaShop/ps_qualityassurance/pull/8/commits/da03ba8a722b9e6ec2da896b72f65667ddc7c0a4

File: ps_qualityassurance.php (hunk 2 above)
Comment (matks) [reply ps_qualityassurance.php]: Fixed in https://github.com/PrestaShop/ps_qualityassurance/pull/8/commits/da03ba8a722b9e6ec2da896b72f65667ddc7c0a4

File: ps_qualityassurance.php (hunk 1 above)
Comment (matks) [reply ps_qualityassurance.php]: Fixed in https://github.com/PrestaShop/ps_qualityassurance/pull/8/commits/da03ba8a722b9e6ec2da896b72f65667ddc7c0a4

**Comment by matks** [comment]
CI is happy 🎉

File: _dev/back/pages/Logs.vue [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          <th>Parameters</th>
+          <th>Output</th>
+          <th>Time</th>
+          <th>Status</th>
+        </tr>
+      </thead>
+      <tbody>
+        <tr v-for="log in logs"
+        :key="log.id"
+        v-bind:class="{ 'table-danger': (log.error === '1') }">
```
Comment (PierreRambaud) [inline _dev/back/pages/Logs.vue:43]:
```suggestion
        :class="{ 'table-danger': (log.error === '1') }">
```

File: _dev/back/pages/Logs.vue [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ *-->
+<template>
+  <div class="panel">
+    <table class="table table-striped"
+    v-for="(logs, requestIdentifier) in logsGroupedByRequest"
+    :key="requestIdentifier">
```
Comment (NeOMakinG) [inline _dev/back/pages/Logs.vue:29]:
```suggestion
    <table 
      class="table table-striped"
      v-for="(logs, requestIdentifier) in logsGroupedByRequest"
      :key="requestIdentifier"
    >
```

File: _dev/back/pages/Logs.vue [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        v-bind:class="{ 'table-danger': (log.error === '1') }">
+          <th scope="row">{{ log.id }}</th>
+          <td>{{ log.hook_name }}</td>
+          <td>{{ log.hook_parameters }}</td>
+          <td>{{ log.output }}</td>
+          <td>{{ log.called_at }}</td>
+          <td>
+            <i
+              class="material-icons"
+            >
```
Comment (NeOMakinG) [inline _dev/back/pages/Logs.vue:52]:
```suggestion
            <i class="material-icons">
```

File: _dev/back/pages/Logs.vue (hunk 10 above)
Comment (matks) [reply _dev/back/pages/Logs.vue]: You broke the build of my PR :trollface:  https://travis-ci.com/github/PrestaShop/ps_qualityassurance/jobs/454388350

File: _dev/back/pages/Logs.vue [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+          <th>Parameters</th>
+          <th>Output</th>
+          <th>Time</th>
+          <th>Status</th>
+        </tr>
+      </thead>
+      <tbody>
+        <tr v-for="log in logs"
+        :key="log.id"
+        :class="{ 'table-danger': (log.error === '1') }">
```
Comment (NeOMakinG) [inline _dev/back/pages/Logs.vue:45]:
```suggestion
        <tr 
          v-for="log in logs"
          :key="log.id"
          :class="{ 'table-danger': (log.error === '1') }"
        >
```

**Review by Progi1984 — APPROVED** [review]

File: ps_qualityassurance.php (hunk 3 above)
Comment (PierreRambaud) [reply ps_qualityassurance.php:130]: C# threads or java threads have no values, sorry bro

Comment (matks) [reply ps_qualityassurance.php:130]: Java is the best ❤️

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Danke
